### PR TITLE
Forbid managed conversations

### PR DESCRIPTION
**Pull request by neongreen**
Now we have an internal endpoint for creating them, and that endpoint is still used for tests. The public endpoint forbids creating them.

### CODE REVIEW

**Comment by neongreen** [comment]
@tiago-loureiro I changed my mind, it's not fair that you review everything

**Comment by neongreen** [comment]
@tiago-loureiro can you clarify how you see two different `ConvTeamInfo`s working? We need to have the `managed` field in the public `ConvTeamInfo` because we want to give out the "deprecated" error message to clients, but if we keep it there and create a new `ConvTeamInfoManaged` or something just for the internal endpoint, then we'd also have to create a new `NewConvManaged` and that would likely be inconvenient

**Comment by tiago-loureiro** [comment]
> then we'd also have to create a new NewConvManaged and that would likely be inconvenient.

Yes, this was the idea; I just think it's safer to rely on types when possible (i.e., on the internal endpoint you look at the type and you know that you can create a managed conversation), similarly to how we allow the creation of [new users](https://github.com/wireapp/wire-server/blob/develop/libs/brig-types/src/Brig/Types/User.hs#L385-L397)

The deprecation could also be expressed in the `FromJSON` instance.

**Comment by neongreen** [comment]
Okay, I will create `newtype NewConvManaged`, remove the `managed` field from `NewConv` and change the instance to `fail` when it encounters `managed: true`. Sounds like a plan?

**Comment by neongreen** [comment]
(@tiago-loureiro)

(not sure if you get notifications here or not)

**Comment by tiago-loureiro** [comment]
> Okay, I will create newtype NewConvManaged, remove the managed field from NewConv and change the instance to fail when it encounters managed: true. Sounds like a plan?

Yup, sounds good to me!

> (@tiago-loureiro)
> (not sure if you get notifications here or not)

Always do :)

**Comment by neongreen** [comment]
@tiago-loureiro after spending two hours on that, I think that we shouldn't remove the `managed` field after all since it's a can of worms -- e.g. `ConvTeamInfo` is also used as a return type for database queries, but we don't want to remove `managed` from the database schema because it will break the logic.

I will make a `newtype NewConvManaged` without removing `managed`, but this is as far as we can go without bringing `-XDataKinds` in, which is doable and gives perfect type safety here -- at the cost of making the code an overengineered mess. I'd rather wait until Liquid Haskell is available.

**Comment by neongreen** [comment]
By the way: what should be done about this type?

```
data TeamConversation = TeamConversation
    { _conversationId      :: ConvId
    , _managedConversation :: Bool
    }
```

Do I understand correctly that it's just a "view" type and the only reason we have a `FromJSON` instance for it is that it's used in Brig-Galley communication? If so, then I'll remove the instance and turn it into a function instead, so that its usage can be tracked more easily.

**Comment by neongreen** [comment]
Also, you said:

> on the internal endpoint you look at the type and you know that you can create a managed conversation

However, I don't see how to do this, since everywhere we have this pattern:

```
createGroupConversation :: UserId ::: ConnId ::: Request ::: JSON -> Galley Response
createGroupConversation (zusr ::: zcon ::: req ::: _) = do
    body :: NewConv <- fromBody req invalidPayload
    ...
```

So, the type doesn't mention `NewConv` at all. Is there something I'm missing? It's be nice to be able to write something like this:

```
createGroupConversation :: UserId ::: ConnId ::: Body NewConv -> Galley Response
```

File: libs/galley-types/src/Galley/Types.hs [hunk 1]
```diff
@@ -174,6 +176,45 @@ data NewConv = NewConv
...
+Those are invariants enforced on the 'FromJSON' level. For convenience, the
+newtype constructors have not been hidden.
+
+The public POST /conversations endpoint only allows unmanaged conversations.
+For creating managed conversations we provide an internal endpoint called
+POST /i/conversations/managed. When an endpoint receives payload
+corresponding to a forbidden conversation type, it throws a JSON parsing
+error, which is not optimal but it doesn't matter since nobody is trying to
+create managed conversations anyway.
+-}
```
[the remark this answers is not in the document]
Comment (fisx) [reply libs/galley-types/src/Galley/Types.hs]: I tried to improve on your `newtype` approach here: a655f80c775eaf22744f08cb511064411eeeeee6, 6e7b10a534c21d955c5ff63327867d2c19cbc97c, but didn't get very far.  If you think this is an improvement I'd like to see how you fix it.  If you don't think this is worth pursuing I agree.

So, all good!

[the remark this answers is not in the document]
Comment (fisx) [reply libs/galley-types/src/Galley/Types.hs]: (to clarify: those two commits are on a different branch.  no need to revert them if we want them gone.)

File: services/galley/src/Galley/API.hs [hunk 2]
```diff
@@ -823,6 +823,12 @@ sitemap = do
         capture "cnv"
         .&. capture "usr"
 
+    post "/i/conversations/managed" (continue internalCreateManagedConversation) $
```
[the remark this answers is not in the document]
Comment (neongreen) [reply services/galley/src/Galley/API.hs]: It's actually consistent-ish: we also have `internalGetMember` and `internalListClients`.

**Pull request merged**